xaitk_saliency/impls/gen_object_detector_blackbox_sal/occlusion_based.py

```diff
@@ -0,0 +1,196 @@
+from xaitk_saliency.interfaces.perturb_image import PerturbImage
+from xaitk_saliency.interfaces.gen_detector_prop_sal import GenerateDetectorProposalSaliency
+from xaitk_saliency.utils.detection import format_detection
+from xaitk_saliency.utils.masking import occlude_image_batch
+
+from smqtk_detection.utils.bbox import AxisAlignedBoundingBox
+from smqtk_detection.interfaces.detect_image_objects import DetectImageObjects
+from smqtk_core.configuration import (
+    from_config_dict,
+    make_default_config,
+    to_config_dict,
+)
+
+C = TypeVar("C", bound="PerturbationOcclusion")
+
+
+class PerturbationOcclusion (GenerateObjectDetectorBlackboxSaliency):
+    """
+    Generator composed of modular perturbation and occlusion-based algorithms.
+
+    This implementation exposes a public attribute `fill`.
+    This may be set to a scalar or sequence value to indicate a color that
+    should be used for filling occluded areas as determined by the given
+    `PerturbImage` implementation.
+    This is a parameter to be set during runtime as this is most often driven
+    by the black-box algorithm used, if at all.
+
+    :param perturber: `PerturbImage` implementation instance for generating
+        occlusion masks.
+    :param generator: `GenerateDetectorProposalSaliency` implementation
+        instance for generating saliency masks given occlusion masks and
+        black-box detector outputs.
+    :param threads: Optional number threads to use to enable parallelism in
+        applying perturbation masks to an input image.
+        If 0, a negative value, or `None`, work will be performed on the
+        main-thread in-line.
+    """
+
+    def __init__(
+        self,
+        perturber: PerturbImage,
+        generator: GenerateDetectorProposalSaliency,
+        threads: Optional[int] = 0
+    ):
+        self._perturber = perturber
+        self._generator = generator
+        self._threads = threads
+        # Optional fill color
+        self.fill: Optional[Union[int, Sequence[int]]] = None
+
+    def _generate(
+        self,
+        ref_image: np.ndarray,
+        bboxes: np.ndarray,
+        scores: np.ndarray,
+        blackbox: DetectImageObjects,
+        objectness: Optional[np.ndarray] = None,
+    ) -> np.ndarray:
+        ref_dets_mat = format_detection(bboxes, scores, objectness)
+
+        pert_masks = self._perturber(ref_image)
+
+        # aggregate to list so that labels can be easily accessed
+        pert_dets = list(blackbox.detect_objects(
+            occlude_image_batch(
+                ref_image,
+                pert_masks,
+                fill=self.fill,
+                threads=self._threads
+            )
+        ))
+
+        # use class labels of first non-empty detection from first image
+        labels = []
+        for img_dets in pert_dets:
+            for det in img_dets:
+                labels = list(det[1].keys())
+                break
+            if len(labels) > 0:
+                break
```

Comment:
This likely OK to have out here, but since the result is only used for the `_dets_to_formatted_mat` function, this logic could be moved into that function. I also suspect you could lazy-populate this static list on your first detection encounter in there instead of a priori flattening all detections for all images.